common/payload/payload.go

```diff
@@ -94,11 +95,28 @@ func MergeMapOfPayload(
 	}
 	res := util.CloneMapNonNil(dst)
 	for k, v := range src {
-		if proto.Equal(v, nilPayload) || proto.Equal(v, emptySlicePayload) {
+		if isEqual(v, nilPayload) || isEqual(v, emptySlicePayload) {
 			delete(res, k)
 		} else {
 			res[k] = v
 		}
 	}
 	return res
 }
+
+// isEqual returns true if both have the same encoding and data.
+// It does not take additional metadata into consideration.
+func isEqual(a, b *commonpb.Payload) bool {
+	return getEncoding(a) == getEncoding(b) && bytes.Equal(a.GetData(), b.GetData())
```

Comment:
FWIW, data equality not required on binary/null, just metadata check (though harmless).

As for empty list, technically `[]`, `[  ]`, ` []` etc are all the same. Probably don't need to change, but just know that semantic equality is not the same as byte equality.

Reply:
That's a good note! I'm adding comments about this. Thanks!